File: app/src/main/java/com/guanhaolin/pearch/adapter/VideoAdapter.java
package com.guanhaolin.pearch.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.guanhaolin.pearch.databinding.VideoViewHolderBinding;
import com.guanhaolin.pearch.model.VideoInfoResponse;
import java.util.ArrayList;
import java.util.List;

public class VideoAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

  public interface OnVideoCellClickListener {
    void onVideoCellClicked(VideoInfoResponse videoInfo);
  }

  private List<VideoInfoResponse> listVideoInfo = new ArrayList<>();

  private VideoAdapter.OnVideoCellClickListener onVideoCellClickListener;

  private final View.OnClickListener onClickListener =
      new View.OnClickListener() {
        @Override
        public void onClick(View view) {
          if (onVideoCellClickListener != null) {
            VideoInfoResponse videoInfo = (VideoInfoResponse) view.getTag();
            if (videoInfo != null) {
              onVideoCellClickListener.onVideoCellClicked(videoInfo);
            }
          }
        }
      };

  public void setOnVideoCellClickListener(
      VideoAdapter.OnVideoCellClickListener onVideoCellClickListener) {
    this.onVideoCellClickListener = onVideoCellClickListener;
  }

  public void update(List<VideoInfoResponse> newVideos) {
    int size = listVideoInfo.size();
    listVideoInfo.clear();
    listVideoInfo.addAll(newVideos);
    notifyItemRangeInserted(size, newVideos.size() - size);
  }

  public void clear() {
    listVideoInfo.clear();
    notifyDataSetChanged();
  }

  @NonNull
  @Override
  public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    VideoViewHolderBinding viewBinding =
        VideoViewHolderBinding.inflate(LayoutInflater.from(parent.getContext()));
    viewBinding.root.setOnClickListener(onClickListener);
    return new VideoViewHolder(viewBinding);
  }

  @Override
  public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
    VideoInfoResponse videoInfo = listVideoInfo.get(position);
    VideoViewHolder viewHolder = (VideoViewHolder) holder;
    viewHolder.bind(videoInfo);
  }

  @Override
  public int getItemCount() {
    return listVideoInfo.size();
  }
}

File: app/src/main/java/com/guanhaolin/pearch/adapter/VideoAdapter.kt
package com.guanhaolin.pearch.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.guanhaolin.pearch.databinding.VideoViewHolderBinding
import com.guanhaolin.pearch.model.VideoInfoResponse

class VideoAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    interface OnVideoCellClickListener {
        fun onVideoCellClicked(videoInfo: VideoInfoResponse?)
    }

    private val listVideoInfo: MutableList<VideoInfoResponse> = ArrayList()

    private var onVideoCellClickListener: OnVideoCellClickListener? = null

    private val onClickListener =
        View.OnClickListener { view ->
            if (onVideoCellClickListener != null) {
                val videoInfo = view.tag as VideoInfoResponse
                if (videoInfo != null) {
                    onVideoCellClickListener!!.onVideoCellClicked(videoInfo)
                }
            }
        }

    fun setOnVideoCellClickListener(
        onVideoCellClickListener: OnVideoCellClickListener?
    ) {
        this.onVideoCellClickListener = onVideoCellClickListener
    }

    fun update(newVideos: List<VideoInfoResponse>) {
        val size = listVideoInfo.size
        listVideoInfo.clear()
        listVideoInfo.addAll(newVideos)
        notifyItemRangeInserted(size, newVideos.size - size)
    }

    fun clear() {
        listVideoInfo.clear()
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val viewBinding =
            VideoViewHolderBinding.inflate(LayoutInflater.from(parent.context))
        viewBinding.root.setOnClickListener(onClickListener)
        return VideoViewHolder(viewBinding)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val videoInfo = listVideoInfo[position]
        val viewHolder = holder as VideoViewHolder
        viewHolder.bind(videoInfo)
    }

    override fun getItemCount(): Int {
        return listVideoInfo.size
    }
}
